File: gh.py
"""Thin wrapper around the `gh` CLI for GitHub API calls."""

import json
import logging
import subprocess
from typing import Any

log = logging.getLogger(__name__)


def _run(args: list[str], check: bool = True) -> str:
    """Run a gh CLI command and return stdout."""
    cmd = ["gh"] + args
    log.debug("Running: %s", " ".join(cmd))
    result = subprocess.run(
        cmd, capture_output=True, text=True, check=check, timeout=120,
        encoding="utf-8", errors="replace",
    )
    if result.returncode != 0:
        log.error("gh stderr: %s", result.stderr.strip())
    return result.stdout.strip()


def api(endpoint: str, method: str = "GET", fields: dict | None = None,
        jq_filter: str | None = None, paginate: bool = False) -> Any:
    """Call the GitHub REST API via `gh api`."""
    args = ["api", endpoint, "--method", method]
    if jq_filter:
        args += ["--jq", jq_filter]
    if paginate:
        args.append("--paginate")
    for k, v in (fields or {}).items():
        args += ["-f", f"{k}={v}"]
    raw = _run(args)
    if not raw:
        return None
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw


def list_open_prs(owner: str, repo: str, author: str,
                  search_query: str = "") -> list[dict]:
    """Search for open PRs matching criteria using gh search flags."""
    args = [
        "search", "prs",
        "--repo", f"{owner}/{repo}",
        "--author", author,
        "--state", "open",
        "--json", "number,title,state,labels,createdAt",
        "--limit", "20",
        "--", search_query,
    ]
    raw = _run(args, check=False)
    if not raw:
        return []
    return json.loads(raw)


def get_pr(owner: str, repo: str, number: int) -> dict:
    return api(f"/repos/{owner}/{repo}/pulls/{number}")


def get_pr_files(owner: str, repo: str, number: int) -> list[dict]:
    return api(f"/repos/{owner}/{repo}/pulls/{number}/files", paginate=True)


def get_pr_reviews(owner: str, repo: str, number: int) -> list[dict]:
    return api(f"/repos/{owner}/{repo}/pulls/{number}/reviews", paginate=True)


def get_pr_review_comments(owner: str, repo: str, number: int) -> list[dict]:
    return api(
        f"/repos/{owner}/{repo}/pulls/{number}/comments", paginate=True
    )


def get_check_runs(owner: str, repo: str, sha: str) -> list[dict]:
    data = api(f"/repos/{owner}/{repo}/commits/{sha}/check-runs",
               jq_filter=".check_runs")
    return data or []


def get_combined_status(owner: str, repo: str, sha: str) -> dict:
    return api(f"/repos/{owner}/{repo}/commits/{sha}/status")


def approve_pr(owner: str, repo: str, number: int, body: str = "") -> dict:
    """Submit an approving review."""
    args = ["pr", "review", str(number), "--approve",
            "--repo", f"{owner}/{repo}"]
    if body:
        args += ["--body", body]
    return _run(args)


def merge_pr(owner: str, repo: str, number: int,
             method: str = "squash") -> str:
    args = ["pr", "merge", str(number), f"--{method}",
            "--repo", f"{owner}/{repo}", "--auto"]
    return _run(args, check=False)


def add_comment(owner: str, repo: str, number: int, body: str) -> dict:
    return api(
        f"/repos/{owner}/{repo}/issues/{number}/comments",
        method="POST",
        fields={"body": body},
    )


def get_authenticated_user() -> str:
    return _run(["auth", "status", "--active", "-t"],
                check=False).split("\n")[0]

File: Gh.cs
using System.Diagnostics;
using System.Text.Json;

namespace UpgradePrAgent;

/// <summary>Thin wrapper around the gh CLI.</summary>
public static class Gh
{
    public static async Task<string> RunAsync(params string[] args)
    {
        var psi = new ProcessStartInfo("gh")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };
        foreach (var a in args) psi.ArgumentList.Add(a);

        using var proc = Process.Start(psi)!;
        var stdout = await proc.StandardOutput.ReadToEndAsync();
        var stderr = await proc.StandardError.ReadToEndAsync();
        await proc.WaitForExitAsync();

        if (proc.ExitCode != 0 && !string.IsNullOrWhiteSpace(stderr))
            Log.Debug($"gh stderr: {stderr.Trim()}");

        return stdout.Trim();
    }

    public static async Task<JsonElement?> ApiAsync(
        string endpoint, string method = "GET",
        Dictionary<string, string>? fields = null,
        string? jqFilter = null, bool paginate = false)
    {
        var args = new List<string> { "api", endpoint, "--method", method };
        if (jqFilter is not null) { args.Add("--jq"); args.Add(jqFilter); }
        if (paginate) args.Add("--paginate");
        if (fields is not null)
        {
            foreach (var (k, v) in fields)
            {
                args.Add("-f"); args.Add($"{k}={v}");
            }
        }

        var raw = await RunAsync([.. args]);
        if (string.IsNullOrWhiteSpace(raw)) return null;

        try { return JsonDocument.Parse(raw).RootElement; }
        catch (JsonException) { return null; }
    }

    public static async Task<List<JsonElement>> ListOpenPrsAsync(
        string owner, string repo, string author, string searchQuery)
    {
        var raw = await RunAsync(
            "search", "prs",
            "--repo", $"{owner}/{repo}",
            "--author", author,
            "--state", "open",
            "--json", "number,title,state,labels,createdAt",
            "--limit", "20",
            "--", searchQuery);

        if (string.IsNullOrWhiteSpace(raw)) return [];

        var doc = JsonDocument.Parse(raw);
        return [.. doc.RootElement.EnumerateArray()];
    }

    public static async Task<JsonElement> GetPrAsync(string owner, string repo, int number) =>
        (await ApiAsync($"/repos/{owner}/{repo}/pulls/{number}"))!.Value;

    public static async Task<List<JsonElement>> GetCheckRunsAsync(string owner, string repo, string sha)
    {
        var el = await ApiAsync($"/repos/{owner}/{repo}/commits/{sha}/check-runs",
            jqFilter: ".check_runs");
        if (el is null) return [];
        return [.. el.Value.EnumerateArray()];
    }

    public static async Task<List<JsonElement>> GetPrReviewsAsync(string owner, string repo, int number)
    {
        var el = await ApiAsync($"/repos/{owner}/{repo}/pulls/{number}/reviews", paginate: true);
        if (el is null) return [];
        return [.. el.Value.EnumerateArray()];
    }

    public static async Task<List<JsonElement>> GetPrReviewCommentsAsync(string owner, string repo, int number)
    {
        var el = await ApiAsync($"/repos/{owner}/{repo}/pulls/{number}/comments", paginate: true);
        if (el is null) return [];
        return [.. el.Value.EnumerateArray()];
    }

    public static async Task<string> ApprovePrAsync(string owner, string repo, int number, string body = "")
    {
        var args = new List<string>
        {
            "pr", "review", number.ToString(), "--approve",
            "--repo", $"{owner}/{repo}"
        };
        if (!string.IsNullOrEmpty(body)) { args.Add("--body"); args.Add(body); }
        return await RunAsync([.. args]);
    }

    public static async Task<string> MergePrAsync(string owner, string repo, int number, string method = "squash") =>
        await RunAsync("pr", "merge", number.ToString(), $"--{method}",
            "--repo", $"{owner}/{repo}", "--auto");

    public static async Task AddCommentAsync(string owner, string repo, int number, string body) =>
        await ApiAsync($"/repos/{owner}/{repo}/issues/{number}/comments",
            method: "POST", fields: new() { ["body"] = body });
}
